### PR TITLE
Pull Request for the 6th Milestone of the Rails project Social Milestone 6: Friendships v2

### PR DESCRIPTION
**Pull request by jose-Abel**
Hello,

We're presenting this Pull Request to be reviewed consisting of the sith milestone for the Rails module, the Social media website project - Milestone 6: Friendships v2. In this milestone, we're requested to implement friendships with 2 rows for mutual friendship, and all the unit and integrations test necessary.

Hoping to get amazing feedback, happy coding!

### CODE REVIEW

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

Hello team! ✋ 
Thanks for submitting this new milestone for a pull request! 👍 
I can see you have only made one commit updating the README file adding the authors! 🤔 
I cannot see any implementation of the basic requirements for this milestone.
Kindly, implement the basic requirements and requests for a new code review once done. 
I know this milestone can be challenging, but at least give it a try! 
Here you got links to [this artcile](https://explainextended.com/2009/03/07/selecting-friends) and [stackoverflow discussion](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table) as a hint to implement the milestone.
Remember to add the appropriate tests for this milestone and modify your controllers and models as necessary.
Good luck with it! 😺

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

### Hello team 😄,

**Your project looks amazing. Great job 💪. There are still a few fixes needed, however, before the project can be approved.**

1). **I will use an example to explain. I have two users, Jen and Ken. They are not friends. When Jen signs in, she can be able to see Ken's posts when she visits Ken's profile. Please fix it so Jen can see Ken's posts, in his profile page, only when they are friends.**

2). **In this example, Jen and Ken are friends. When I look for all of Jen's friends in the console, I get two rows of Ken's details. See screenshot below and look into it.**

friends

3). **Add a method which I can use to query for all of a user's friend requests sent and not yet confirmed. Example, `user.pending_requests`.**

4). **Add a method which I can use to query for all of a user's friend requests received and not yet confirmed. Example, `user.friend_requests`.**

5). **Add a method which I can use to query all of a user's posts and those of the user's friends. Example, `user.friends_and_own_posts`.**

6). **Add tests for your `helpers` methods. For any methods that are not being used outside a module, you can declare them private and not write tests for them.**

7). **I got the error below after I ran `RSpec`. I tried using the instructions [here](https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver#requirements) but it still throws an error. You can add in your readme how one can solve this error when using your code.**
rspec


### Happy coding 🥂.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Hello, team! ✋ 
A nice evolution in this project, I can see you have worked really hard on it! 👍 
Well done implementing the changes requested by the previous TSE! 👏 
However, we need to make a few changes more to approve this project! 💪 
Kindly check my comments below 👇  apply the changes and request a new code review ⏩ 
Good luck with it! 😃 
## Mandatory changes
- The error running RSpec is still there, kindly fix it
- We need to fix our user model since we are not getting the expected behaviour, the friends_request and the pending request methods are not working properly, kindly fix them.
- In your user model you should add a couple of new associations. making the relationships between models and columns who creates the friendship, ie.`has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
`

**Comment by lekegitrepo** [comment]
Hi, I was unable to run migration because of the username and password you add to the database configuration. So, I'm going to mark this as invalid.

File: app/models/user.rb [hunk 1]
```diff
@@ -13,34 +11,31 @@ class User < ApplicationRecord
...
-  def friend?(user)
-    friends.include?(user)
-  end
+  # def friend_requests
+  #   inverted_friendships.map { |fs| fs.user unless fs.confirmed }.compact
+  # end
+
+  # def pending_requests
+  #   friendships.map { |fs| fs.friend unless fs.confirmed }.compact
+  # end
```
Comment (ghost) [inline app/models/user.rb]: Remove this commented code.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 

## Hello team 😃,

**You have done a great job implementing the requested changes 💪. A few things, however,**

1). **You have not written tests for the added methods in `user.rb` They include: `:pending_friends` and `:friend_requests`.**

2). **I have followed the instructions you gave to try and solve the rspec error and now I am getting an error as shown in the screenshot below. If the error persists, I suggest you reach out to one of the full-time TSEs for help. I will provide their slack details below. One more thing, your have misspelt `rspec` in your instruction manual. Please correct that to avoid confusion.**

rspec

3). **Check one more in the comment below.**

4). **Full-time TSE contacts on Slack**

- Sarah - Sarah Uryelah Silva Chamorro

- Bola - Bola Ahmed Buari

- Maru - Maria Eugenia Quemé Peña

### Happy coding 🥂 

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by SpaYco — APPROVED** [review]
# Project Approved ✔️ 

thank you for making the required changes, your project is now approved!

**_P.S: the problem that the previous TSEs faced was due to the chrome browser (and default path) being needed to process your test, please for future tests try to rely on gems that doesn't rely on these things._**

### please don't forget to merge to master, good luck with the next projects!


Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Pull request merged**